datafusion-examples/examples/file_stream_provider.rs

```diff
@@ -0,0 +1,185 @@
+            if interval < broken_pipe_timeout {
+                thread::sleep(Duration::from_millis(100));
+                continue;
+            }
+        }
+        return exec_err!("{}", e);
+    }
+    Ok(())
+}
+
+fn create_writing_thread(
+    file_path: PathBuf,
+    maybe_header: Option<String>,
+    lines: Vec<String>,
+    waiting_lock: Arc<AtomicBool>,
+    wait_until: usize,
+) -> JoinHandle<()> {
+    // Timeout for a long period of BrokenPipe error
+    let broken_pipe_timeout = Duration::from_secs(10);
+    let sa = file_path.clone();
+    // Spawn a new thread to write to the FIFO file
+    #[allow(clippy::disallowed_methods)] // spawn allowed only in tests
+    spawn_blocking(move || {
+        let file = OpenOptions::new().write(true).open(sa).unwrap();
+        // Reference time to use when deciding to fail the test
+        let execution_start = Instant::now();
+        if let Some(header) = maybe_header {
+            write_to_fifo(&file, &header, execution_start, broken_pipe_timeout).unwrap();
+        }
+        for (cnt, line) in lines.iter().enumerate() {
+            while waiting_lock.load(Ordering::SeqCst) && cnt > wait_until {
+                thread::sleep(Duration::from_millis(50));
+            }
+            write_to_fifo(&file, line, execution_start, broken_pipe_timeout).unwrap();
+        }
+        drop(file);
+    })
+}
+
+/// This example demonstrates a scanning against an Arrow data source (JSON) and
+/// fetching results
+#[tokio::main]
+async fn main() -> Result<()> {
+    // Create session context
+    let config = SessionConfig::new()
+        .with_batch_size(TEST_BATCH_SIZE)
+        .with_collect_statistics(false)
+        .with_target_partitions(1);
+    let ctx = SessionContext::new_with_config(config);
+    let tmp_dir = TempDir::new()?;
+    let fifo_path = create_fifo_file(&tmp_dir, "fifo_unbounded.csv")?;
+
+    let mut tasks: Vec<JoinHandle<()>> = vec![];
+    let waiting = Arc::new(AtomicBool::new(true));
+
+    let data_iter = 0..TEST_DATA_SIZE;
+    let lines = data_iter
+        .map(|i| format!("{},{}\n", i, i + 1))
+        .collect::<Vec<_>>();
+    // Create writing threads for the left and right FIFO files
+    tasks.push(create_writing_thread(
+        fifo_path.clone(),
+        Some("a1,a2\n".to_owned()),
+        lines.clone(),
+        waiting.clone(),
+        TEST_DATA_SIZE,
+    ));
+
+    // Create schema
+    let schema = Arc::new(Schema::new(vec![
+        Field::new("a1", DataType::UInt32, false),
+        Field::new("a2", DataType::UInt32, false),
+    ]));
+
+    // Specify the ordering:
+    let order = vec![vec![datafusion_expr::col("a1").sort(true, false)]];
+
+    let provider = fifo_table(schema.clone(), fifo_path, order.clone());
+    ctx.register_table("fifo", provider)?;
+
+    let df = ctx.sql("SELECT * FROM fifo").await.unwrap();
+    let mut stream = df.execute_stream().await.unwrap();
+
+    futures::future::join_all(tasks).await;
```

Comment:
I wonder if you need to join all here (it kind of makes the streaming nature of this example less clear) -- could you potentially do after you read the batches?

Reply:
Will try that out